msrestazure/azure_active_directory.py

```diff
@@ -657,29 +658,108 @@ class MSIAuthentication(BasicTokenAuthentication):
     - resource (str): Alternative authentication resource, default
       is 'https://management.core.windows.net/'.
 
-    :param int port: MSI local port if VM/VMSS context (ignored otherwise)
-
     .. versionadded:: 0.4.14
     """
 
     def __init__(self, port=50342, **kwargs):
         super(MSIAuthentication, self).__init__(None)
 
+        if port != 50342:
+            warnings.warn("The 'port' argument is no longer used, and will be removed in a future release", DeprecationWarning)
         self.port = port
+
         self.msi_conf = {k:v for k,v in kwargs.items() if k in ["client_id", "object_id", "msi_res_id"]}
 
         self.cloud_environment = kwargs.get('cloud_environment', AZURE_PUBLIC_CLOUD)
         self.resource = kwargs.get('resource', self.cloud_environment.endpoints.active_directory_resource_id)
 
-    def set_token(self):
         if _is_app_service():
             if self.msi_conf:
                 raise AuthenticationError("User Assigned Entity is not available on WebApp yet.")
+        elif "MSI_ENDPOINT" not in os.environ:
+            # Use IMDS if no MSI_ENDPOINT
+            self._vm_msi = _ImdsTokenProvider(self.resource, self.msi_conf)
+
+    def set_token(self):
+        if _is_app_service():
             self.scheme, _, self.token = get_msi_token_webapp(self.resource)
-        else:
+        elif "MSI_ENDPOINT" in os.environ:
             self.scheme, _, self.token = get_msi_token(self.resource, self.port, self.msi_conf)
+        else:
+            token_entry = self._vm_msi.get_token()
+            self.scheme, self.token = token_entry['token_type'], token_entry
 
     def signed_session(self):
         # Token cache is handled by the VM extension, call each time to avoid expiration
         self.set_token()
         return super(MSIAuthentication, self).signed_session()
+
+
+class _ImdsTokenProvider(object):
+    """A help class handling token acquisitions through Azure IMDS plugin.
+    """
+
+    def __init__(self, resource, msi_conf=None):
+        self.identity_type, self.identity_id = None, None
+        if msi_conf:
+            if len(msi_conf.keys()) > 1:
+                raise ValueError('"client_id", "object_id", "msi_res_id" are mutually exclusive')
+            elif len(msi_conf.keys()) == 1:
+                self.identity_type, self.identity_id = next(iter(msi_conf.items()))
+        # default to system assigned identity on an empty configuration object
+
+        self.cache = {}
+        self.resource = resource
+
+    def get_token(self):
```

Comment:
Can we surface this in the `MSIAuthentication`? CLI could use it to support `az account get-access-token`, which is a popular command

Reply:
This should be available by doing:
```python
creds.set_token()
return creds.token
```
Thoughts?

Reply:
Can we just expose a `get_token` with these 2 lines? Otherwise it looks more like a hack for client apps to sniff out the privates, thoughts?